### PR TITLE
[Harness] If empty nodes are found, do not crash.

### PR DESCRIPTION
**Pull request by mandel-macaque**
Looks like we have cases in which we get empty nodes when listing
devices, which makes the parsing of the nodes fail and crash.

Move the code to a diff method, try to parse the data, if there is an
exception (missing data) return null. If null is returned, do not add
it.

Fixes: https://github.com/xamarin/xamarin-macios/issues/8160

### CODE REVIEW

**Comment by mandel-macaque** [comment]
Why we get empty nodes from apple is unclear, might be a bug in their side :/

File: tests/xharness/Xharness.Tests/Hardware/Tests/DevicesTest.cs [hunk 1]
```diff
@@ -72,7 +72,7 @@ public void LoadAsyncProcessErrorTest (bool timeout)
 
 			MlaunchArgument sdkRootArg = passedArguments.Where (a => a is SdkRootArgument).FirstOrDefault();
 			Assert.IsNotNull (sdkRootArg, "sdk arg missing");
-			AssertArgumentValue (sdkRootArg, sdkPath, "sdk arg wrong");
+			AssertArgumentValue (sdkRootArg, $"--sdkroot {sdkPath}", "sdk arg wrong");
```
Comment (spouliot) [inline tests/xharness/Xharness.Tests/Hardware/Tests/DevicesTest.cs]: is this related to the change ?

Comment (mandel-macaque) [reply tests/xharness/Xharness.Tests/Hardware/Tests/DevicesTest.cs]: No, but the tests failed, there was a commit that should have updated this.

File: tests/xharness/Hardware/Devices.cs [hunk 2]
```diff
@@ -62,6 +62,31 @@ Task ILoadAsync.LoadAsync (ILog log, bool include_locked, bool force)
...
+					BuildVersion = deviceNone.SelectSingleNode ("BuildVersion")?.InnerText,
+					ProductVersion = deviceNone.SelectSingleNode ("ProductVersion")?.InnerText,
+					ProductType = deviceNone.SelectSingleNode ("ProductType")?.InnerText,
+					InterfaceType = deviceNone.SelectSingleNode ("InterfaceType")?.InnerText,
+					IsUsableForDebugging = usable == null ? (bool?) null : ((bool?) (usable == "True")),
+				};
+				bool.TryParse (deviceNone.SelectSingleNode ("IsLocked")?.InnerText, out var locked);
+				device.IsLocked = locked;
+				return device;
+			} catch {
```
Comment (rolfbjarne) [inline tests/xharness/Hardware/Devices.cs]: A solution that doesn't involve throwing and catching exceptions is much better. One example is that exceptions tracing can be used to figure out real problems.

File: tests/xharness/Xharness.Tests/Hardware/Tests/DevicesTest.cs (hunk 1 above)
Comment (spouliot) [reply tests/xharness/Xharness.Tests/Hardware/Tests/DevicesTest.cs]: That should be a separate PR or, minimally, mentioning in the commit message.
Also it should link to the original commit and check why this was merged if not working.

Comment (mandel-macaque) [reply tests/xharness/Xharness.Tests/Hardware/Tests/DevicesTest.cs]: @spouliot issue with these tests, they are not in our CI, xharness is not running its own tests, might be a good idea to add that.

File: tests/xharness/Hardware/Devices.cs (hunk 2 above)
Comment (mandel-macaque) [reply tests/xharness/Hardware/Devices.cs]: Fixed, the TryParse should work correctly, if the DeviceClass cannot be parsed, return null.

File: tests/xharness/Xharness.Tests/Hardware/Tests/DevicesTest.cs (hunk 1 above)
Comment (spouliot) [reply tests/xharness/Xharness.Tests/Hardware/Tests/DevicesTest.cs]: Tests that don't run don't exists :(
Let's move that unrelated fix into a PR that enable the tests :)

Comment (premun) [reply tests/xharness/Xharness.Tests/Hardware/Tests/DevicesTest.cs]: I opened a PR to fix these and couple of others that are failing: https://github.com/xamarin/xamarin-macios/pull/8208

Comment (mandel-macaque) [reply tests/xharness/Xharness.Tests/Hardware/Tests/DevicesTest.cs]: @premun thx!!!! I'll revert that change then!
